src/server/bitops_family.cc

```diff
@@ -685,6 +748,90 @@ OpResult<std::size_t> CountBitsForValue(const OpArgs& op_args, std::string_view
   }
 }
 
+// Returns the bit position (where MSB is 0, LSB is 7) of the leftmost bit that
+// equals `value` in `byte`. Returns 8 if not found.
+std::size_t GetFirstBitWithValueInByte(uint8_t byte, bool value) {
+  if (value) {
+    return absl::countl_zero(byte);
+  } else {
+    return absl::countl_one(byte);
+  }
+}
+
+int64_t FindFirstBitWithValueAsBit(std::string_view result_value, bool bit_value, int64_t start,
+                                   int64_t end) {
+  for (int64_t i = start; i <= end; ++i) {
+    if (static_cast<size_t>(GetByteIndex(i)) >= result_value.size()) {
+      break;
+    }
+    const uint8_t current_byte = GetByteValue(result_value, i);
+    bool current_bit = CheckBitStatus(current_byte, GetNormalizedBitIndex(i));
+    if (current_bit != bit_value) {
+      continue;
+    }
+
+    return i;
+  }
+
+  return -1;
+}
+
+int64_t FindFirstBitWithValueAsByte(std::string_view result_value, bool bit_value, int64_t start,
+                                    int64_t end) {
+  for (int64_t i = start; i <= end; ++i) {
+    if (static_cast<size_t>(i) >= result_value.size()) {
+      break;
+    }
+    const uint8_t current_byte = result_value[i];
+    const uint8_t kNotFoundByte = bit_value ? 0 : std::numeric_limits<uint8_t>::max();
+    if (current_byte == kNotFoundByte) {
+      continue;
+    }
+
+    return i * OFFSET_FACTOR + GetFirstBitWithValueInByte(current_byte, bit_value);
+  }
+
+  return -1;
+}
+
+OpResult<int64_t> FindFirstBitWithValue(const OpArgs& op_args, std::string_view key, bool bit_value,
+                                        int64_t start, int64_t end, bool as_bit) {
+  OpResult<std::string> result = ReadValue(op_args.db_cntx, key, op_args.shard);
+
+  std::string_view result_value;
+  if (result) {  // non-existent keys are treated as empty strings, per Redis
+    result_value = result.value();
+  }
+
+  int64_t size = result_value.size();
+  if (as_bit) {
+    size *= OFFSET_FACTOR;
+  }
+
+  int64_t normalized_start = NormalizedOffset(size, start);
+  int64_t normalized_end = NormalizedOffset(size, end);
+  if (normalized_start > normalized_end) {
+    return -1;  // Return -1 for negative ranges, per Redis
+  }
+
+  int64_t position;
+  if (as_bit) {
+    position =
+        FindFirstBitWithValueAsBit(result_value, bit_value, normalized_start, normalized_end);
+  } else {
+    position =
+        FindFirstBitWithValueAsByte(result_value, bit_value, normalized_start, normalized_end);
+  }
+
+  if (position == -1 && !bit_value && static_cast<size_t>(start) < result_value.size() &&
+      end == std::numeric_limits<int64_t>::max()) {
+    // Not-found return value is compatible with Redis, but is subtle.
```

Comment:
The comment is a bit enigmatic to me :) 

If returning -1 is compatible with Redis I am not sure I agree with the logic, or with the cognitive burden on the reader this edge-case yields.

Reply:
In this specific case, Redis returns (and documents that it returns) the size, in bits, of the value.
I agree, this edge case is super weird and has a potentially significant burden on users. But that's the Redis API, and Dragonfly aims to be compatible, so we must mimic this weird API :(
I slightly modified the comment, please lmk if that's not good enough or if you have other suggestions.